src/tls13.c

```diff
@@ -8637,9 +8637,18 @@ int wolfSSL_connect_TLSv13(WOLFSSL* ssl)
              * fragment, fragOffset is zero again, and the state can be
              * advanced. */
             if (ssl->fragOffset == 0) {
-                ssl->options.connectState++;
-                WOLFSSL_MSG("connect state: "
-                            "Advanced from last buffered fragment send");
+                /* Only increment from states in which we send data */
+                if (ssl->options.connectState == CONNECT_BEGIN ||
+                    ssl->options.connectState == HELLO_AGAIN ||
+                    ssl->options.connectState == FIRST_REPLY_DONE ||
+                    ssl->options.connectState == FIRST_REPLY_FIRST ||
+                    ssl->options.connectState == FIRST_REPLY_SECOND ||
+                    ssl->options.connectState == FIRST_REPLY_THIRD ||
+                    ssl->options.connectState == FIRST_REPLY_FOURTH) {
+                    ssl->options.connectState++;
+                    WOLFSSL_MSG("connect state: "
```

Comment:
If we could redesign the implementation I suppose it would be better to set flags to do that but unfortunately that would be a big rewrite. This is just the minimal way to fix the issue.